### PR TITLE
Mute/unmute video properly.

### PR DESCRIPTION
**Pull request by aforge**
Now muting turns the camera off instead of sending blank frames.

### CODE REVIEW

File: src/widgets/call-panel.jsx [hunk 1]
```diff
@@ -139,6 +149,16 @@ class CallPanel extends React.PureComponent {
     }
   }
 
+  // Creates an empty video track placeholder.
+  emptyVideoTrack() {
+    const width = 640;
+    const height = 480;
+    let canvas = Object.assign(document.createElement("canvas"), {width, height});
```
Comment (or-else) [inline src/widgets/call-panel.jsx:156]: const canvas ...

File: src/widgets/call-panel.jsx [hunk 2]
```diff
@@ -139,6 +149,16 @@ class CallPanel extends React.PureComponent {
...
+    canvas.getContext('2d').fillRect(0, 0, width, height);
+    let stream = canvas.captureStream();
```
Comment (or-else) [inline src/widgets/call-panel.jsx:158]: const stream ...
You should probably use explicit frame rate of 0, i.e.
`const stream = canvas.captureStream(0);`

File: src/widgets/call-panel.jsx [hunk 3]
```diff
@@ -155,6 +175,12 @@ class CallPanel extends React.PureComponent {
     // Start local video.
     navigator.mediaDevices.getUserMedia(this.localStreamConstraints)
       .then(stream => {
+        if (!this.localStreamConstraints.video) {
+          // Starting an audio-only call. Create a dummy video track
+          // (so video can be enabled during the call if the user desires).
+          const dummy = this.emptyVideoTrack();
+          stream.addTrack(dummy);
```
Comment (or-else) [inline src/widgets/call-panel.jsx:182]: Maybe just `stream.addTrack(this.emptyVideoTrack());` ?

File: src/widgets/call-panel.jsx [hunk 4]
```diff
@@ -308,6 +334,13 @@ class CallPanel extends React.PureComponent {
   handleTrackEvent(event) {
     // Remote video becomes available.
     this.remoteRef.current.srcObject = event.streams[0];
+
+    if (event.track.kind == 'video') {
+      // Redraw the screen when remote video stream state changes.
+      event.track.onended = _ => { this.forceUpdate(); };
+      event.track.onmute = () => { this.forceUpdate(); };
```
Comment (or-else) [inline src/widgets/call-panel.jsx:341]: Consistently `_` or `()` ?

Just `event.track.onended = _ => this.forceUpdate();`

File: src/widgets/call-panel.jsx [hunk 5]
```diff
@@ -308,6 +334,13 @@ class CallPanel extends React.PureComponent {
...
+      event.track.onunmute = () => { this.forceUpdate(); };
```
Comment (or-else) [inline src/widgets/call-panel.jsx:342]: `_` or `()` ?

File: src/widgets/call-panel.jsx [hunk 6]
```diff
@@ -388,38 +433,74 @@ class CallPanel extends React.PureComponent {
...
+        stream.removeTrack(track);
+        // Add this track to the existing local stream.
+        this.state.localStream.addTrack(track);
+        return sender.replaceTrack(track);
+      })
+      .then(_ => {
+        this.localRef.current.srcObject = this.state.localStream;
+      })
+      .catch(this.handleGetUserMediaError)
+      .finally(() => { this.forceUpdate() }); // Make sure we redraw the mute/unmute icons (e.g. camera -> camera_off).
```
Comment (or-else) [inline src/widgets/call-panel.jsx:465]: Consistently `()` or `_`

`.finally(_ => this.forceUpdate());`

**Comment by or-else** [comment]
Start audio call, press unmute-mute video button a few times, get a crash:

`call-panel.jsx:360 Error opening your camera and/or microphone: Cannot read properties of undefined (reading 'getSenders')`

  | handleGetUserMediaError | @ | call-panel.jsx:360
-- | -- | -- | --
  | Promise.catch (async) |   |  
  | unmuteVideo | @ | call-panel.jsx:464
  | handleToggleCameraClick | @ | call-panel.jsx:473

**Comment by or-else** [comment]
Muting video freezes the picture but does not clear the screen to default (gray background with the avatar in the middle).

File: src/widgets/call-panel.jsx (hunk 1 above)
Comment (aforge) [reply src/widgets/call-panel.jsx]: Done.

File: src/widgets/call-panel.jsx (hunk 2 above)
Comment (aforge) [reply src/widgets/call-panel.jsx]: Done.

File: src/widgets/call-panel.jsx (hunk 3 above)
Comment (aforge) [reply src/widgets/call-panel.jsx]: Done.

File: src/widgets/call-panel.jsx (hunk 4 above)
Comment (aforge) [reply src/widgets/call-panel.jsx]: Done.

File: src/widgets/call-panel.jsx (hunk 5 above)
Comment (aforge) [reply src/widgets/call-panel.jsx]: Done.

File: src/widgets/call-panel.jsx (hunk 6 above)
Comment (aforge) [reply src/widgets/call-panel.jsx]: Done.

**Comment by aforge** [comment]
> Start a call, press unmute-mute video button a few times, get a crash:
> 
> `call-panel.jsx:360 Error opening your camera and/or microphone: Cannot read properties of undefined (reading 'getSenders')`
> 
>  	handleGetUserMediaError	@	call-panel.jsx:360
>  	Promise.catch (async)	 	 
>  	unmuteVideo	@	call-panel.jsx:464
>  	handleToggleCameraClick	@	call-panel.jsx:473

Fixed.

**Comment by aforge** [comment]
> Muting video freezes the picture then the screen start flipping between two states:
> 
> Screenshot 2023-01-09 at 17 43 02
> 
> Screenshot 2023-01-09 at 17 38 17
> 
> I'm calling between FF and Chrome.
> 
> Audio must be enabled, otherwise it stays in the freeze-frame state.

Does the video freeze on FF? Or on both Chrome and FF?

**Comment by or-else** [comment]
> > Start a call, press unmute-mute video button a few times, get a crash:

> Fixed

Now crashes in a different place (tested in FF):

Error opening your camera and/or microphone: pc is undefined [call-panel.jsx:361:16](webpack://tinode-webapp/src/widgets/call-panel.jsx)
    handleGetUserMediaError call-panel.jsx:361
    unmuteVideo call-panel.jsx:466
    handleToggleCameraClick call-panel.jsx:480
    React 23
    <anonymous> index.js:46
    <anonymous> index.dev.js:21663
    <anonymous> index.dev.js:21665

**Comment by aforge** [comment]
> 

Are you seeing any errors in either's console?

**Comment by or-else** [comment]
> Are you seeing any errors in either's console?

Nothing in the console, just flashes once every 3-5 seconds on Chrome. Stays frozen-frame on FF.

**Comment by or-else** [comment]
Chrome still crashes with

`Error opening your camera and/or microphone: Cannot set properties of null (setting 'srcObject')`

**Comment by or-else** [comment]
> Are you seeing any errors in either's console?

You are probably changing state (this.setState()) in response to some WebRTC event.

**Comment by or-else** [comment]
I just sent you a video of flashing through GPhotos

**Comment by or-else** [comment]
still crashing when i press video button in on the caller's end in chrome:

`call-panel.jsx:362 Error opening your camera and/or microphone: Cannot read properties of undefined (reading 'getSenders')`

**Comment by or-else** [comment]
| handleGetUserMediaError | @ | call-panel.jsx:362
-- | -- | -- | --
  | Promise.catch (async) |   |  
  | unmuteVideo | @ | call-panel.jsx:467
  | handleToggleCameraClick | @ | call-panel.jsx:481

**Pull request merged**